### PR TITLE
fix(frontend): refine error msg on materialized views(#3178)

### PR DESCRIPTION
**Pull request by infdahai**
Signed-off-by: clundro <clusternlh@gmail.com>

## What's changed and what's your intention?

Add an unit test `drop_materialized_view_using_drop_source`  in `handler/drop_source.rs` .

Add some judges on `materialized view` before function `get_source_by_name`(with the error msg `source not found`).

## Checklist

- [√] I have written necessary docs and comments
- [√] I have added necessary unit tests and integration tests
- [√] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)

issue: https://github.com/singularity-data/risingwave/issues/3178

### CODE REVIEW

File: src/frontend/src/binder/relation/table_or_source.rs [hunk 1]
```diff
@@ -161,9 +161,26 @@ impl Binder {
 
     pub(crate) fn bind_table_source(&mut self, name: ObjectName) -> Result<BoundTableSource> {
         let (schema_name, source_name) = Self::resolve_table_name(name)?;
-        let source = self
-            .catalog
-            .get_source_by_name(&self.db_name, &schema_name, &source_name)?;
+        let source =
```
Comment (neverchanje) [inline src/frontend/src/binder/relation/table_or_source.rs:164]: When will we reach this logic? Can you add tests as well?

File: src/frontend/src/catalog/root_catalog.rs [hunk 2]
```diff
@@ -183,7 +183,19 @@ impl Catalog {
                     Err(CatalogError::Duplicated("table", relation_name.to_string()).into())
                 }
                 risingwave_pb::stream_plan::source_node::SourceType::Source => {
-                    Err(CatalogError::Duplicated("source", relation_name.to_string()).into())
+                    let table = self
+                        .get_table_by_name(db_name, schema_name, relation_name)
```
Comment (neverchanje) [inline src/frontend/src/catalog/root_catalog.rs:187]: pls add tests for this

CREATE TABLE t
CREATE SOURCE s ()

File: src/frontend/src/binder/relation/table_or_source.rs [hunk 3]
```diff
@@ -161,9 +161,26 @@ impl Binder {
...
+            {
+                Ok(it) => it,
+                Err(err) => {
+                    if let Ok(table) =
+                        self.catalog
+                            .get_table_by_name(&self.db_name, &schema_name, &source_name)
+                            &&table.associated_source_id().is_none()
+                    {
+                            return Err(RwError::from(ErrorCode::InvalidInputSyntax(
+                                "Materized view doesn't support this syntax.".to_owned(),
```
Comment (BugenZhao) [inline src/frontend/src/binder/relation/table_or_source.rs:177]: We may follow the error message with Postgres.

image

File: src/frontend/src/handler/drop_source.rs [hunk 4]
```diff
@@ -18,17 +18,38 @@ use risingwave_pb::stream_plan::source_node::SourceType;
...
+
+    let source: SourceCatalog;
+    {
+        source = match catalog_reader.read_guard().get_source_by_name(
+            session.database(),
+            &schema_name,
+            &source_name,
+        ) {
+            Ok(it) => it.clone(),
+            Err(err) => {
```
Comment (BugenZhao) [inline src/frontend/src/handler/drop_source.rs:38]: May use `or_else(..)` to simplify this.

File: src/frontend/src/handler/drop_source.rs [hunk 5]
```diff
@@ -114,4 +135,23 @@ mod tests {
                 .to_string()
         );
     }
+
+    #[tokio::test]
+    async fn test_drop_materialized_view_using_drop_source() {
```
Comment (BugenZhao) [inline src/frontend/src/handler/drop_source.rs:140]: We may write these tests as planner tests under `src/frontend/test_runner/tests/testdata` so they can be more intuitive and easy to update.

**Review by BugenZhao — COMMENTED** [review]
Thanks for your contribution.

**Comment by skyzh** [comment]
Please sign CLA :)

**Comment by infdahai** [comment]
```sh
dev=> CREATE TABLE t;
CREATE_TABLE
dev=> create materialized view mv as select * from t;
CREATE_MATERIALIZED_VIEW
dev=> insert into mv values (233);
ERROR:  Invalid input syntax: cannot change materialized view mv
```
I change the original error msg to  `Invalid input syntax: cannot change materialized view mv`.

**Comment by infdahai** [comment]
```yaml
# - sql:  |
#     create table t;
#     create source t with ( 'connector' = 'datagen' ) row format json;
# - sql:  |
#     create source t with ( 'connector' = 'datagen' ) row format json;
#     create table t;
# - sql:  |
#     create materialized source t (v1 int) with ( 'connector' = 'datagen' )  row format json;
#     create source t with ( 'connector' = 'datagen' ) row format json;
# - sql:
#     create table s;
#     create materialized view mv as select * from s;
#     drop source mv;
```
The above test cases can produce `Catalog` error msg. And I havn't found the corresponding solution.

**Comment by BugenZhao** [comment]
> ```yaml
> ```yaml
>  - sql:  |
>      create table t;
>      create source t with ( 'connector' = 'datagen' ) row format json;
> #  error msg: failed basic_query.yaml 
> #   failed on case #22 (id: <none>): Catalog error: table with name t exists: Catalog error: table with name t exists
>  - sql:  |
>      create source t with ( 'connector' = 'datagen' ) row format json;
>      create table t;
> #  error msg: failed basic_query.yaml 
> #   failed on case #22 (id: <none>): Catalog error: source with name t exists: Catalog error: source with name t exists
>  - sql:  |
>      create materialized source t (v1 int) with ( 'connector' = 'datagen' )  row format json;
>      create source t with ( 'connector' = 'datagen' ) row format json;
> #  error msg: failed basic_query.yaml 
> #   failed on case #22 (id: <none>): Catalog error: materialized source with name t exists: Catalog error: materialized source with name t exists
>  - sql:
>      create table s;
>      create materialized view mv as select * from s;
>      drop source mv;
> #  error msg: failed basic_query.yaml 
> #   failed on case #22 (id: <none>): Invalid input syntax: Use `DROP MATERIALIZED VIEW` to drop a materialized view.: Invalid input syntax: Use `DROP MATERIALIZED VIEW` to drop a materialized view.
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> The above test cases can produce Catalog error msg and I havn't found the corresponding solution. So I comment out these sql codes.
> ```

Do you mean we cannot capture these error in the test runner? cc @skyzh

**Comment by infdahai** [comment]
> Do you mean we cannot capture these error in the test runner? cc @skyzh

I add the  `catalog_err` part to the  test runner and then the command `./risedev do-apply-planner-test` outputs `Diff applied!`.

File: src/frontend/src/binder/relation/table_or_source.rs [hunk 6]
```diff
@@ -161,9 +161,20 @@ impl Binder {
...
             .catalog
-            .get_source_by_name(&self.db_name, &schema_name, &source_name)?;
+            .get_source_by_name(&self.db_name, &schema_name, &source_name)
+            .map_err(|err| {
+                if let Ok(table) = self.catalog.get_table_by_name(&self.db_name, &schema_name, &source_name)
+                && table.associated_source_id().is_none(){
+                    return RwError::from(ErrorCode::InvalidInputSyntax(
+                        format!("cannot change materialized view {}",source_name)
+                    ));
+                }
```
Comment (liurenjie1024) [inline src/frontend/src/binder/relation/table_or_source.rs:174]: I think we should not report error here, `bind_table_source` is not only used in create materiazed view.

**Comment by infdahai** [comment]
I report this error for changing mv operations, not creating. So this error will throwed in catalog error. And creating mv error occurs  in bind error. 

@liurenjie1024  I search the  `bind_table_source` and find it's invoked on src/insert.rs, src/delete.rs and src/update.rs .

So `bind_table_source` just is used by `insert`,`delete` and `update` option.

**Comment by liurenjie1024** [comment]
> 



> I report this error for changing mv operations, not creating. So this error will be thrown in binder error. And creating mv error occurs in catalog error.
> 
> @liurenjie1024 I search the `bind_table_source` and find it's invoked on binder/insert.rs, binder/delete.rs and binder/update.rs .
> 
> So `bind_table_source` just is used by `insert`,`delete` and `update` option.

I don't think we should do this check in binder just because it's only used by those. We should explicity disable changing mv in handlers, which makes code clear and more readable.

**Comment by infdahai** [comment]
> > 
> 
> > I report this error for changing mv operations, not creating. So this error will be thrown in binder error. And creating mv error occurs in catalog error.
> > @liurenjie1024 I search the `bind_table_source` and find it's invoked on binder/insert.rs, binder/delete.rs and binder/update.rs .
> > So `bind_table_source` just is used by `insert`,`delete` and `update` option.
> 
> I don't think we should do this check in binder just because it's only used by those. We should explicity disable changing mv in handlers, which makes code clear and more readable.

I understand that. And I move this code to concrete ` insert,delete and update` error .

**Comment by fuyufjh** [comment]
Any updates?

**Comment by xxchan** [comment]
Kindly ping @infdahai. Are you still intersted in this PR? If not, I will help resolve conflicts and refine it later.

**Pull request closed without merging**